### PR TITLE
Support custom memory limit

### PR DESCRIPTION
**Pull request by jens1o**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -54,5 +54,17 @@
     "mz": "^2.4.0",
     "semver": "^5.3.0",
     "vscode-languageclient": "^3.0.3"
+  },
+  "contributes": {
+    "configuration": {
+      "title": "PHP-Intellisense",
+      "properties": {
+        "phpintellisense.memoryLimit": {
```
Comment (felixfbecker) [inline package.json]: The configuration section used is `php.*`

Comment (jens1o) [inline package.json]: Yes, but I thought it's kinda worrying for the user. That's why I added this.

**Comment by jens1o** [comment]
like this?

File: src/extension.ts [hunk 2]
```diff
@@ -51,7 +64,7 @@ export async function activate(context: vscode.ExtensionContext): Promise<void>
     const serverOptions = () => new Promise<ChildProcess | StreamInfo>((resolve, reject) => {
         function spawnServer(...args: string[]): ChildProcess {
             // The server is implemented in PHP
-            args.unshift(context.asAbsolutePath(path.join('vendor', 'felixfbecker', 'language-server', 'bin', 'php-language-server.php')));
+            args.unshift(context.asAbsolutePath(path.join('vendor', 'felixfbecker', 'language-server', 'bin', 'php-language-server.php') + ' --memory-limit=' + memoryLimit));
```
Comment (felixfbecker) [inline src/extension.ts]: The way you are unshifting this here it ends up not as an arg but as a part of the executable file.

Comment (jens1o) [inline src/extension.ts]: yes. That's working as designed. How should I execute it else?

Comment (felixfbecker) [inline src/extension.ts]: push it to the args array

Comment (jens1o) [inline src/extension.ts]: Okay.

File: src/extension.ts [hunk 3]
```diff
@@ -12,6 +12,19 @@ export async function activate(context: vscode.ExtensionContext): Promise<void>
     const conf = vscode.workspace.getConfiguration('php');
     const executablePath = conf.get<string>('executablePath') || 'php';
 
+    const memoryLimit = conf.get<string>('memoryLimit') || '-1';
+
+    if (memoryLimit !== '-1' && !memoryLimit.match('~^(\d+)(K|M|G|)$~')) {
```
Comment (felixfbecker) [inline src/extension.ts]:
```ts
!/^\d+(?:K|M|G|)$/.exec(memoryLimit)
```

File: package.json [hunk 4]
```diff
@@ -54,5 +54,16 @@
...
     "semver": "^5.3.0",
     "vscode-languageclient": "^3.0.3"
+  },
+  "contributes": {
+    "configuration": {
+      "properties": {
+        "php.memoryLimit": {
+          "type": "string",
+          "default": "-1",
+          "description": "The memory limit of the php language server. [Number][K|M|G]. Use '-1' to allow unlimited use of the RAM(default)."
```
Comment (felixfbecker) [inline package.json]: I think you can add a regexp to json schema.

File: package.json [hunk 5]
```diff
@@ -54,5 +54,17 @@
...
     "vscode-languageclient": "^3.0.3"
+  },
+  "contributes": {
+    "configuration": {
+      "properties": {
+        "php.memoryLimit": {
+          "type": "string",
+          "default": "-1",
+          "description": "The memory limit of the php language server. [Number][K|M|G]. Use '-1' to allow unlimited use of the RAM(default).",
+          "schema": "!/^\d+(?:K|M|G|)$/"
```
Comment (felixfbecker) [inline package.json]: you need to escape the backslash.
why is this in `schema` and not in `pattern`? why the slashes and exclamation mark?
 https://spacetelescope.github.io/understanding-json-schema/reference/regular_expressions.html

File: src/extension.ts [hunk 6]
```diff
@@ -52,6 +65,7 @@ export async function activate(context: vscode.ExtensionContext): Promise<void>
         function spawnServer(...args: string[]): ChildProcess {
             // The server is implemented in PHP
             args.unshift(context.asAbsolutePath(path.join('vendor', 'felixfbecker', 'language-server', 'bin', 'php-language-server.php')));
+            args.push(' --memory-limit=' + memoryLimit)
```
Comment (felixfbecker) [inline src/extension.ts]: remove the space at the beginning

File: package.json (hunk 5 above)
Comment (jens1o) [inline package.json]: Because I want to match the start to the end.

Comment (felixfbecker) [inline package.json]: yes, that's what ^ and $ are for

File: package.json [hunk 7]
```diff
@@ -54,5 +54,17 @@
...
     "vscode-languageclient": "^3.0.3"
+  },
+  "contributes": {
+    "configuration": {
+      "properties": {
+        "php.memoryLimit": {
+          "type": "string",
+          "default": "-1",
+          "description": "The memory limit of the php language server. [Number][K|M|G]. Use '-1' to allow unlimited use of the RAM(default).",
+          "pattern": "^\\d+(?:K|M|G|)$"
```
Comment (felixfbecker) [inline package.json]: there is a superfluous `|` in the group

File: src/extension.ts [hunk 8]
```diff
@@ -12,6 +12,19 @@ export async function activate(context: vscode.ExtensionContext): Promise<void>
     const conf = vscode.workspace.getConfiguration('php');
     const executablePath = conf.get<string>('executablePath') || 'php';
 
+    const memoryLimit = conf.get<string>('memoryLimit') || '-1';
+
+    if (memoryLimit !== '-1' && !/^\d+(?:K|M|G|)$/.exec(memoryLimit)) {
```
Comment (felixfbecker) [inline src/extension.ts]: here too

File: package.json (hunk 7 above)
Comment (jens1o) [inline package.json]: Won't it break when there is pure bit-data provided?

File: src/extension.ts (hunk 8 above)
Comment (jens1o) [inline src/extension.ts]: see previous question.

File: package.json (hunk 7 above)
Comment (felixfbecker) [inline package.json]: Oh, okay, now I see what you wanted to do. You can write this much simpler as
```regexp
^\d+[KMG]?$
```

Comment (jens1o) [inline package.json]: Thanks!

**Pull request merged**